beeline.go

```diff
@@ -97,65 +102,146 @@ func Init(config Config) {
+// Additionally, these fields are packaged up and passed along to downstream
+// processes if they are also using a beeline. This function is good for adding
+// context that is better scoped to the request than this specific unit of work,
+// eg user IDs, globally relevant feature flags, errors, etc. Fields added here
+// are prefixed with `global.`
+func AddFieldToTrace(ctx context.Context, key string, val interface{}) {
+	namespacedKey := fmt.Sprintf("global.%s", key)
+	trace := internal.GetTraceFromContext(ctx)
+	if trace != nil {
+		trace.AddField(namespacedKey, val)
 	}
-	return nil
+}
+
+// HasTrace returns true if there is a trace in the current context
+func HasTrace(ctx context.Context) bool {
+	trace := internal.GetTraceFromContext(ctx)
+	return trace != nil
+}
+
+// StartTraceWithIDs lets you start a trace with a specific set of IDs - it is
+// used when you've received the IDs from another source (eg incoming HTTP
+// headers). If you don't care what the IDs are, you may use either this or
+// StartSpan to start a trace. You cannot change the trace IDs after a trace has
+// begun. StartTraceWithIDs also starts a span (the root span) for this trace.
+// You should call FinishSpan to close the span (and trace) started by this
+// function.
+func StartTraceWithIDs(ctx context.Context, traceID, parentID, name string) context.Context {
+	ctx, _ = internal.StartTraceWithIDs(ctx, traceID, parentID, name)
+	return ctx
+}
+
+// StartSpan lets you start a new span as a child of an already instrumented
+// handler. If there isn't an existing wrapped handler in the context when this
+// is called, it will start a new trace. Spans automatically get a `duration_ms`
+// field when they are ended; you should not explicitly set the duration unless
+// you want to override it. The name argument will be the primary way the span
+// is identified in the trace view within Honeycomb. You must use the returned
+// context to ensure attributes are added to the correct span.
+func StartSpan(ctx context.Context, name string) context.Context {
+	ctx, _ = internal.StartSpan(ctx, name)
+	return ctx
+}
+
+// TODO do we need to provide some way to extract the trace / span from this
+// context and shove it in to another? I suspect strange things would happen to
+// any downstream services that want to use the context for timeouts or
+// cancellation if the request finished (which triggers the original context
+// getting cancelled, which is most likely inappropriate for an async span).
+
+// StartAsyncSpan is different from StartSpan in that when finishing a trace, it
+// does not get automatically sent. When finishing an async span it gets sent
+// immediately.
+func StartAsyncSpan(ctx context.Context, name string) context.Context {
+	ctx, _ = internal.StartAsyncSpan(ctx, name)
+	return ctx
+}
+
+// FinishSpan finishes the currently active span. It should only be called for
+// spans created with StartTraceWithIDs, StartSpan, or StartAsyncSpan. Use the
```

Comment:
It's a little unclear how to follow this advice ("`FinishSpan`... should only be called for spans created with ... `StartSpan`"). `FinishSpan` doesn't take a `span` parameter, so how am I supposed to know (besides diligence and code inspection) which span I'm finishing? The wording sort of implies that a) spans are things I can get my hands on (qv @emfree's comment), and b) there might be other ways to create spans besides `StartTraceWithIDs`, `StartSpan`, or `StartAsyncSpan`, but I should mistrust those other ways, or use some other way of finishing spans created thus.

Maybe clearer guidance would be "only call this after calling `StartSpan` etc, and don't call it more than once", or similar?